### DISCUSSION TITLE
Cannot create a simple jest test

### DISCUSSION THREAD

**Discussion by kapitanluffy** [vercel/next.js#75455] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #22 SMP Tue Jan 10 18:39:00 UTC 2023
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: 7.1.0
    Relevant packages:
      next: 13.3.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/lively-leftpad-ket1gt

### To Reproduce

Run tests with `npx jest`

### Describe the Bug

When I use jest with `nextJest` the tests fail

### Expected Behavior

Tests should pass

**Comment by JesseKoldewijn** [comment]
euhm... not to just to any conclusions but it more likely seems like your test is just incorrect rather than this being an issue on Next.js 's part.

**Comment by kapitanluffy** [comment]
> euhm... not to just to any conclusions but it more likely seems like your test is just incorrect rather than this being an issue on Next.js 's part.

In the `jest.config.js`, comment line 13 and uncomment line 16, then run `npx jest`. Let me know if the test passes or not.

**Comment by JesseKoldewijn** [comment]
On my end I was able to run the test but the test itself seems incorrect. What error do you get?

**Comment by kapitanluffy** [comment]
> On my end I was able to run the test but the test itsepf seems incorrect. What error do you get?

Do the opposite of my previous comment. Uncomment line 13 and comment line 16, then run npx jest.

Also, what should be the correct test? Does that mean the documentation provided by jest is wrong?

https://jestjs.io/docs/jest-object#jestspyonobject-methodname

**Comment by JesseKoldewijn** [comment]
I'm mainly curious about what error you're getting since I'm able to run the test but it throws errors on the test itself and not on the result it should give.

**Comment by JesseKoldewijn** [comment]
From the errors I saw and about 2 min of googling it seemed like it had trouble detecting the ran function called "bar" by the way you imported it. Made a small change on it which you can see in the codesandbox below. Hope this helped ya (works on both config variants you mentioned)

https://codesandbox.io/p/sandbox/funny-villani-0x6j4t

**Comment by JesseKoldewijn** [comment]
Nontheless, this is not really an issue it next.js itself but in the way jest behaves with curtain import strats.

**Comment by kapitanluffy** [comment]
> I'm mainly curious about what error you're getting since I'm able to run the test but it throws errors on the test itself and not on the result it should give.

The error is "Cannot redefine property bar". I will send a screenshot once I get back.

**Comment by kapitanluffy** [comment]
> Nontheless, this is not really an issue it next.js itself but in the way jest behaves with curtain import strats. 

Sure. but assuming you checked the codesandbox link I provided. Why is it when I remove "createJestConfig", the test works if it is a problem with how the test is written?

**Comment by JesseKoldewijn** [comment]
> > Nontheless, this is not really an issue it next.js itself but in the way jest behaves with curtain import strats.
> 
> Sure. but assuming you checked the codesandbox link I provided. Why is it when I remove "createJestConfig", the test works if it is a problem with how the test is written?

That's something I don't really get either, initially the issue was in the test itself not being setup in such a way that jest understands it.

**Comment by JesseKoldewijn** [comment]
it might have something to do with the non-standard file/folder structure you're using which might trip jest if its setup in a way it doesn't expect by default. But not sure on that one. There's already multiple issues on jest itself where people also ran into this same or close'ish issue which would indeed conclude that this is not specific to next.js

**Comment by kapitanluffy** [comment]
After further reading, it seems like swc won't be fixing this

https://github.com/swc-project/swc/issues/5059

https://github.com/swc-project/swc/issues/5205


For anyone who will stumble upon this, your tests are fine. Next.js is using SWC as their compiler, which treats ts-jest transpiled tests wrong.

You can fix this by doing one of the following:

- If you are using ts-jest, do not use `nextJest`. Just export your jest config normally
- If you want to still use ts-jest with `nextJest`, use [jest_workaround](https://github.com/swc-project/swc/discussions/5151#discussioncomment-3149154) swc plugin. Add it in your next.config.js
- Use vitest instead

**Comment by JesseKoldewijn** [comment]
> After further reading, it seems like swc won't be fixing this
> 
> [swc-project/swc#5059](https://github.com/swc-project/swc/issues/5059)
> 
> [swc-project/swc#5205](https://github.com/swc-project/swc/issues/5205)
> 
> For anyone who will stumble upon this, your tests are fine. Next.js is using SWC as their compiler, which treats ts-jest transpiled tests wrong.
> 
> You can fix this by doing one of the following:
> 
> * If you are using ts-jest, do not use `nextJest`. Just export your jest config normally
> * If you want to still use ts-jest with `nextJest`, use [jest_workaround](https://github.com/swc-project/swc/discussions/5151#discussioncomment-3149154) swc plugin. Add it in your next.config.js
> * Use vitest instead

Or write your tests a little different. I changed a really minor bit on your test which matches up with non next.js related jest issues and got it to work even with ts-jest and nextJest... 

Since there is a fairly simple solution and there also are lots of non-next.js users reporting the same issue on the jest repo etc. I feel like it's kind of a misconception to see this either as a swc/next.js issue tbh.👍 

The same behavior was also reported on esbuild-jest quite a while ago, see:
https://github.com/aelbore/esbuild-jest/issues/26#issuecomment-893763840
So this would mean in my eyes that there is no real coalition between swc (to be specific in this case) and the issue at hand.

**Comment by kapitanluffy** [comment]
> > After further reading, it seems like swc won't be fixing this
> > 
> > [swc-project/swc#5059](https://github.com/swc-project/swc/issues/5059)
> > 
> > [swc-project/swc#5205](https://github.com/swc-project/swc/issues/5205)
> > 
> > For anyone who will stumble upon this, your tests are fine. Next.js is using SWC as their compiler, which treats ts-jest transpiled tests wrong.
> > 
> > You can fix this by doing one of the following:
> > 
> > * If you are using ts-jest, do not use `nextJest`. Just export your jest config normally
> > * If you want to still use ts-jest with `nextJest`, use [jest_workaround](https://github.com/swc-project/swc/discussions/5151#discussioncomment-3149154) swc plugin. Add it in your next.config.js
> > * Use vitest instead
> 
> Or write your tests a little different. I changed a really minor bit on your test which matches up with non next.js related jest issues and got it to work even with ts-jest and nextJest... 
> 
> Since there is a fairly simple solution and there also are lots of non-next.js users reporting the same issue on the jest repo etc. I feel like it's kind of a misconception to see this either as a swc/next.js issue tbh.👍 

Can you provide me a link to the sample test you made on how to spy a function?

**Comment by JesseKoldewijn** [comment]
> From the errors I saw and about 2 min of googling it seemed like it had trouble detecting the ran function called "bar" by the way you imported it. Made a small change on it which you can see in the codesandbox below. Hope this helped ya (works on both config variants you mentioned)
> 
> https://codesandbox.io/p/sandbox/funny-villani-0x6j4t

@kapitanluffy as mentioned in a previous comment, I've forked your codesandbox and included this fix which will make the test actually run.

**Comment by kapitanluffy** [comment]
I see. I will try it when I get back. This is how I imported it before 🤔

**Comment by JesseKoldewijn** [comment]
> I see. I will try it when I get back. This is how I imported it before 🤔

Let me know if this would have fixed your issue! And if by manual replication you can't get it to work you can always just rip the codesandbox into your vscode and run a diff overtop of your own project to see what might have changed 😉

**Comment by kapitanluffy** [comment]
> > I see. I will try it when I get back. This is how I imported it before 🤔
> 
> Let me know if this would have fixed your issue! And if by manual replication you can't get it to work you can always just rip the codesandbox into your vscode and run a diff overtop of your own project to see what might have changed 😉

The test is wrong. You did not check whether `bar` was called. In line 13 of foo.test.ts, you just wrote `expect(barMock)`

```ts
import { jest, describe, it, expect } from "@jest/globals";
import { foo } from "./foo";
import { bar } from "./bar";

const Mods = { bar };

describe("Foo", () => {
  it("Foo", async () => {
    const barMock = jest.spyOn(Mods, "bar");
    const result = await foo({ msg: "hello" });

    expect(result.msg).toBe("hello");
    expect(barMock);
  });
});
```

**Comment by JesseKoldewijn** [comment]
> 

True but that wasn't the point, the issue itself was in the barMock declaration since that's the line its threw up on. What I would recommend is doing a fly-by on the jest docs and maybe use mock instead of spyOn in this case. It might also help to use a standard folder structure in the project (not sure if this would solve any issues in this case tho but that's not really the point in this case).

**Comment by kapitanluffy** [comment]
> > 
> 
> True but that wasn't the point, the issue itself was in the barMock declaration since that's the line its threw up on. What I would recommend is doing a fly-by on the jest docs and maybe use mock instead of spyOn in this case. It might also help to use a standard folder structure in the project (not sure if this would solve any issues in this case tho but that's not really the point in this case).

I already tried mock instead of using spyOn. Also, the folder structure is from the nextjs template in codesandbox so I don't think there's anything wrong with the folder structure here.

Sure, it is throwing up the error `Cannot redefine property: bar` in the `barMock` declaration but why does it only happen if my config is like this:

```js
const nextJest = require("next/jest");
const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: "./",
});

const config = {
  preset: "ts-jest",
  testEnvironment: "node",
};

module.exports = createJestConfig(config);
```

But the tests pass without errors if my config is like this:

```js

const config = {
  preset: "ts-jest",
  testEnvironment: "node",
};

module.exports = config;
```

You can try it yourself. The test is simple. If you can check whether `bar` is being called at least once inside `foo` using jest regardless of what you use (spy/mock) and without using DI, I'll close this issue. Feel free to use a "standard" folder structure if that helps.

**Comment by JesseKoldewijn** [comment]
Just to clarify, what I meant by a "standard" folder structure is either one of the two bellow:

1. SrcDir
```
/
/src
-> /app
-> /components
```

2. Root
```
/
/app
/components
```

Easiest way to get a folder struct like either of these two options is just by using creat-next-app and choosing any of the following options:
"Do you want to use the src folder"
Yes = srcDir
No = rootDir
